executor/merge_join.go

```diff
@@ -0,0 +1,502 @@
+		} else {
+			rb.peekedRow = curRow
+			return rowCache, nil
+		}
+	}
+}
+
+// Close implements the Executor Close interface.
+func (e *MergeJoinExec) Close() error {
+	e.prepared = false
+	e.cursor = 0
+	e.outputBuf = nil
+
+	lErr := e.leftRowBlock.reader.Close()
+	if lErr != nil {
+		e.rightRowBlock.reader.Close()
+		return errors.Trace(lErr)
+	}
+	rErr := e.rightRowBlock.reader.Close()
+	if rErr != nil {
+		return errors.Trace(rErr)
+	}
+
+	return nil
+}
+
+// Schema implements the Executor Schema interface.
+func (e *MergeJoinExec) Schema() *expression.Schema {
+	return e.schema
+}
+
+func compareKeys(stmtCtx *variable.StatementContext,
+	leftRow *Row, leftKeys []*expression.Column,
+	rightRow *Row, rightKeys []*expression.Column) (int, error) {
+	for i, leftKey := range leftKeys {
+		lVal, err := leftKey.Eval(leftRow.Data)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		rVal, err := rightKeys[i].Eval(rightRow.Data)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		ret, err := lVal.CompareDatum(stmtCtx, rVal)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		if ret != 0 {
+			return ret, nil
+		}
+	}
+	return 0, nil
+}
+
+func (e *MergeJoinExec) outputJoinRow(leftRow *Row, rightRow *Row, preserve bool) error {
+	var joinedRow *Row
+	if e.flipSide {
+		joinedRow = makeJoinRow(rightRow, leftRow)
+	} else {
+		joinedRow = makeJoinRow(leftRow, rightRow)
+	}
+	if e.otherFilter != nil {
+		matched, err := expression.EvalBool(e.otherFilter, joinedRow.Data, e.ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if matched {
+			e.outputBuf = append(e.outputBuf, joinedRow)
+			return nil
+		} else if preserve {
+			if e.flipSide {
+				joinedRow = makeJoinRow(e.defaultRightRow, leftRow)
+			} else {
+				joinedRow = makeJoinRow(leftRow, e.defaultRightRow)
+			}
+		} else {
+			return nil
+		}
+	}
+	e.outputBuf = append(e.outputBuf, joinedRow)
+	return nil
+}
+
+func (e *MergeJoinExec) tryOutputLeftRows() error {
+	if e.preserveLeft {
+		for _, lRow := range e.leftRows {
+			err := e.outputJoinRow(lRow, e.defaultRightRow, true)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (e *MergeJoinExec) computeCrossProduct() error {
+	var err error
+	for _, lRow := range e.leftRows {
+		// make up for outer join since we ignored single table conditions previously
+		if e.leftFilter != nil {
+			matched, err := expression.EvalBool(e.leftFilter, lRow.Data, e.ctx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if !matched {
+				// as all right join converted to left, we only output left side if no match and continue
+				if e.preserveLeft {
+					err := e.outputJoinRow(lRow, e.defaultRightRow, true)
+					if err != nil {
+						return errors.Trace(err)
+					}
+				}
+				continue
+			}
+		}
+		// Do the real cross product calculation
+		initInnerLen := len(e.outputBuf)
+		for _, rRow := range e.rightRows {
+			err = e.outputJoinRow(lRow, rRow, false)
```

Comment:
Preserve should be `e.preseveLeft`.
Then we don't need this argument in `outputJoinRow`, use `e.preseveLeft` will do.